### ISSUE TITLE
tf.reduce_mean on int32/uint32 overflows under jit_compile=True and returns a negative mean of positive inputs

### ISSUE THREAD

**Issue by lackjava-creator** [tensorflow/tensorflow#125978]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

2.15.0; tf-nightly 2.22.0-dev20260823

### Custom code

Yes

### OS platform and distribution

macOS 26.0, Apple Silicon (arm64)

### Python version

3.10 (TF 2.15.0); 3.12.10 (tf-nightly)

### GPU model and memory

N/A — CPU only

### Current behavior?

`tf.reduce_mean` on an `int32` tensor returns a different—and mathematically wrong—result under `tf.function(jit_compile=True)` than in eager execution when the sum of the reduced elements overflows the element type.

```python
a = tf.constant([[2**31 - 1, 2**31 - 1]], dtype=tf.int32)
tf.reduce_mean(a, axis=1)
# eager                    -> [2147483647]
# tf.function(jit_compile) -> [-1]
```

One operator and two elements are enough. The JIT result is consistent with accumulating in the element dtype, wrapping, and then dividing: it returns a **negative mean for strictly positive inputs**.

### Expected behavior

The compiled result should agree with eager execution and return `[2147483647]`.

### Scope

The full matrix below was measured on TensorFlow 2.15.0, CPU. The minimal `int32` and `uint32` cases, as well as the duplication example below, were also confirmed on `tf-nightly 2.22.0-dev20260823` (Git version `v1.12.1-143961-g64fbb70af66`).

| condition | result |
|---|---|
| `int32`, `uint32` | **wrong** |
| `int8`, `int16`, `int64`, `uint8`, `uint16`, `uint64` | agrees with eager |
| `reduce_mean` | **wrong** |
| `reduce_sum`, `reduce_max`, `reduce_prod` | agree with eager |
| `float32`, `float64` | agree with eager |
| `jit_compile=True` | **wrong** |
| autoclustering (`tf.config.optimizer.set_jit(True)`) | agrees with eager |
| values that do not overflow | agrees with eager |
| `axis=0`, `axis=1`, whole-tensor reduce | all wrong |
| repeated runs | deterministic, 5/5 |

For the overflowing `int64` control, both paths return `-1`; it therefore does not produce an eager-versus-compiled discrepancy.

A further consequence is that `reduce_mean` under JIT is not invariant under data duplication. These expressions have the same mathematical mean, but only the second has an overflowing internal sum:

```python
half = (2**31 - 1) // 2
v = tf.constant([[half, half + 1]], dtype=tf.int32)
tf.reduce_mean(v, axis=1)                          # jit and eager: 1073741823
tf.reduce_mean(tf.concat([v, v], axis=1), axis=1) # jit: 0, eager: 1073741823
```

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf


def mean1(t):
    return tf.reduce_mean(t, axis=1)


def check(label, fn, *args):
    eager = fn(*args).numpy()
    compiled = tf.function(
        fn, jit_compile=True, autograph=False
    )(*args).numpy()
    print(
        f"{label:40s} eager={eager}  jit={compiled}  "
        f"{'ok' if np.array_equal(eager, compiled) else 'MISMATCH'}"
    )


print("tensorflow", tf.__version__)

M = 2**31 - 1
a = tf.constant(np.array([[M, M]], dtype=np.int32))

check("int32 [[MAX, MAX]] reduce_mean", mean1, a)
check(
    "int32 [[10, 20]] reduce_mean",
    mean1,
    tf.constant(np.array([[10, 20]], dtype=np.int32)),
)
check(
    "int32 reduce_sum (control)",
    lambda t: tf.reduce_sum(t, axis=1),
    a,
)

for dt in (
    "int8", "int16", "int32", "int64",
    "uint8", "uint16", "uint32", "uint64",
):
    i = np.iinfo(dt)
    check(
        f"{dt} reduce_mean",
        mean1,
        tf.constant(np.array([[i.max, i.max]], dtype=dt)),
    )
```

### Relevant log output

```shell
TensorFlow Nightly:

tensorflow 2.22.0-dev20260823
Git version v1.12.1-143961-g64fbb70af66
int32:  eager=[2147483647]  jit=[-1]          MISMATCH
uint32: eager=[4294967295]  jit=[2147483647]  MISMATCH
dup-base:   eager=[1073741823] jit=[1073741823] ok
dup-concat: eager=[1073741823] jit=[0]          MISMATCH

TensorFlow 2.15.0:

int32 [[MAX, MAX]] reduce_mean           eager=[2147483647]  jit=[-1]  MISMATCH
int32 [[10, 20]] reduce_mean             eager=[15]          jit=[15] ok
int32 reduce_sum (control)               eager=[-2]          jit=[-2] ok
int8 reduce_mean                         eager=[127]         jit=[127] ok
int16 reduce_mean                        eager=[32767]       jit=[32767] ok
int32 reduce_mean                        eager=[2147483647]  jit=[-1] MISMATCH
int64 reduce_mean                        eager=[-1]          jit=[-1] ok
uint8 reduce_mean                        eager=[255]         jit=[255] ok
uint16 reduce_mean                       eager=[65535]       jit=[65535] ok
uint32 reduce_mean                       eager=[4294967295]  jit=[2147483647] MISMATCH
uint64 reduce_mean                       eager=[9223372036854775807] jit=[9223372036854775807] ok
```

**Comment by vishwakt** [comment]
Reproduced on macOS arm64 CPU with TF 2.21.0, matching your matrix exactly. I've opened PR #126012 with a proposed fix.

The cause is the accumulator type chosen for Mean. `MeanOp` in tensorflow/compiler/tf2xla/kernels/reduction_ops.cc accumulates in `XlaHelpers::SumAccumulationType(input_type)` and then divides, and that helper (tensorflow/compiler/tf2xla/xla_helpers.cc line 140) widens only the narrow integer types:

    // Upcast small integer types to 32 bit to avoid overflow.
    if (dtype == DT_INT8 || dtype == DT_INT16) return DT_INT32;
    if (dtype == DT_UINT8 || dtype == DT_UINT16) return DT_UINT32;
    return dtype;

int32 and uint32 fall through unwidened, so the sum wraps before the division. That reproduces your table row for row: int8 and int16 widen to int32 and agree, uint8 and uint16 widen to uint32 and agree, int64 and uint64 have nothing wider and wrap in eager too so they agree, and int32 and uint32 are the two that diverge.

reduce_sum agreeing at -2 is consistent with this: Sum returns the wrapped sum either way because the result is converted back at the end, so only Mean turns a wrapped accumulator into a wrong answer. The PR therefore gives Mean its own accumulation type that widens the 32 bit integer types, rather than changing SumAccumulationType itself, which feeds around sixteen other call sites that do not need it.

Worth noting for anyone else looking at this: there are three copies of this kind of helper, and the one above is the one that is actually live for reduce_mean. The MLIR reduction path uses a separate `GetAccumulationType` (mlir/tf2xla/transforms/legalize_tf.cc line 761) that widens no integer types at all, so if that path were running, int8 would return -1 rather than the 127 you observed.

**Comment by Kayyuri** [comment]
Hi @lackjava-creator, Thank you for reporting this issue!
A [PR](https://github.com/tensorflow/tensorflow/pull/126012) has been submitted to fix this issue, and it will be resolved once merged.

Thank you to @vishwakt for working on the fix contributions.
